src/bin/pg_autoctl/lock_utils.c

```diff
@@ -0,0 +1,283 @@
+/*
+ * src/bin/pg_autoctl/lock_utils.c
+ *   Implementations of utility functions for inter-process locking
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <signal.h>
+#include <sys/types.h>
+#include <sys/ipc.h>
+#include <sys/sem.h>
+#include <unistd.h>
+
+#include "defaults.h"
+#include "file_utils.h"
+#include "env_utils.h"
+#include "lock_utils.h"
+#include "log.h"
+
+
+/*
+ * See man semctl(2)
+ */
+#if !defined(__APPLE__)
+union semun
+{
+	int val;
+	struct semid_ds *buf;
+	unsigned short *array;
+};
+#endif
+
+/*
+ * semaphore_init creates or opens a named semaphore for the current process.
+ *
+ * We use the environment variable PG_AUTOCTL_SERVICE to signal when a process
+ * is a child process of the main pg_autoctl supervisor so that we are able to
+ * initialise our locking strategy before parsing the command line. After all,
+ * we might have to log some output during the parsing itself.
+ */
+bool
+semaphore_init(Semaphore *semaphore)
+{
+	if (env_exists(PG_AUTOCTL_SERVICE))
+	{
+		return semaphore_open(semaphore);
+	}
+	else
+	{
+		return semaphore_create(semaphore);
+	}
+}
+
+
+/*
+ * semaphore_finish closes or unlinks given semaphore.
+ */
+bool
+semaphore_finish(Semaphore *semaphore)
+{
+	if (env_exists(PG_AUTOCTL_SERVICE))
+	{
+		/* there's no semaphore closing protocol in SysV */
+		return true;
+	}
+	else
+	{
+		return semaphore_unlink(semaphore);
+	}
+}
+
+
+/*
+ * semaphore_create creates a new semaphore with the value 1.
+ */
+bool
+semaphore_create(Semaphore *semaphore)
+{
+	union semun semun;
+	pid_t pid = getpid();
+
+	semaphore->pid = pid;
+	semaphore->semId = semget(pid, 1, IPC_CREAT | IPC_EXCL | 0600);
+
+	if (semaphore->semId < 0)
+	{
+		fformat(stderr, "Failed to create semaphore: %m\n");
```

Comment:
We certainly can't call `log_` functions from within the semaphore acquire/release lock functions of course. The creation of the semaphore happens before `log_set_lock` so we should be able to use the log functions there. The unlink function might be called from `atexit` in the future, so let's not consider we have a clean situation with the locking. So that's one function call that may change, doing that.